File: LAB 2/BRESEN_ALGO.m
clc; clear; close all;

X1 = 2; Y1 = 3;
X2 = 12; Y2 = 8;

DX = abs(X2 - X1);
DY = abs(Y2 - Y1);

if X1 < X2
    X_INC = 1;
else
    X_INC = -1;
end

if Y1 < Y2
    Y_INC = 1;
else
    Y_INC = -1;
end

X = X1;
Y = Y1;
ERROR = 2 * DY - DX;

figure;
hold on;
grid on;
axis equal;
title('Bresenham''s Line Algorithm');
xlabel('X-axis'); ylabel('Y-axis');

plot(X, Y, 'ks', 'MarkerSize', 8, 'MarkerFaceColor', 'k');

for I = 1:DX
    if ERROR >= 0
        Y = Y + Y_INC;
        ERROR = ERROR - 2 * DX;
    end
    X = X + X_INC;
    ERROR = ERROR + 2 * DY;
    plot(X, Y, 'ks', 'MarkerSize', 8, 'MarkerFaceColor', 'k');
end

plot([X1 X2], [Y1 Y2], 'r--');

File: LAB 2/BRESEN_ALGO.py
import matplotlib.pyplot as plt

X1, Y1 = 2, 3
X2, Y2 = 12, 8

DX = abs(X2 - X1)
DY = abs(Y2 - Y1)

X_INC = 1 if X1 < X2 else -1
Y_INC = 1 if Y1 < Y2 else -1

X, Y = X1, Y1
ERROR = 2 * DY - DX

plt.figure()
plt.grid(True)
plt.axis('equal')
plt.title('Bresenham\'s Line Algorithm')
plt.xlabel('X-axis')
plt.ylabel('Y-axis')

X_VALS, Y_VALS = [X], [Y]

for I in range(DX):
    if ERROR >= 0:
        Y += Y_INC
        ERROR -= 2 * DX
    X += X_INC
    ERROR += 2 * DY
    X_VALS.append(X)
    Y_VALS.append(Y)

plt.plot(X_VALS, Y_VALS, 'ks', markersize=8, markerfacecolor='k')
plt.plot([X1, X2], [Y1, Y2], 'r--')
plt.show()
